### ISSUE TITLE
Would you plan to release your train code on raw image dataset(SIDD)?

### ISSUE THREAD

**Issue by madfff** [TaoHuang2018/Neighbor2Neighbor#7]

**Comment by madfff** [comment]
I still hava a little question. Is it right to generate subimages on the packed 4-channel raw images?
Thank you for your patience.

**Comment by TaoHuang2018** [comment]
yes, on the packed 4-channel raw images.

**Comment by zejinwang** [comment]
> I still hava a little question. Is it right to generate subimages on the packed 4-channel raw images? Thank you for your patience.

Hello, I have also encountered a problem that cannot be reproduced on SIDD raw-RGB as high as 51.06dB. I implemented it directly on the source code provided by the author, and psnr can only reach 46.7dB. This is my code:

for epoch in range(1, opt.n_epoch + 1):
    cnt = 0

    for param_group in optimizer.param_groups:
        current_lr = param_group['lr']
    print("LearningRate of Epoch {} = {}".format(epoch, current_lr))

    network.train()
    for iteration, noisy in enumerate(TrainingLoader):
        st = time.time()
        noisy = noisy.cuda()
        # pack raw data
        noisy = space_to_depth(noisy, 2)

        optimizer.zero_grad()

        mask1, mask2 = generate_mask_pair(noisy)
        noisy_sub1 = generate_subimages(noisy, mask1)
        noisy_sub2 = generate_subimages(noisy, mask2)
        with torch.no_grad():
            noisy_denoised = network(noisy)
        noisy_sub1_denoised = generate_subimages(noisy_denoised, mask1)
        noisy_sub2_denoised = generate_subimages(noisy_denoised, mask2)

        noisy_output = network(noisy_sub1)
        noisy_target = noisy_sub2
        Lambda = epoch / opt.n_epoch * opt.increase_ratio
        diff = noisy_output - noisy_target
        exp_diff = noisy_sub1_denoised - noisy_sub2_denoised

        loss1 = torch.mean(diff**2)
        loss2 = Lambda * torch.mean((diff - exp_diff)**2)
        loss_all = opt.Lambda1 * loss1 + opt.Lambda2 * loss2

        loss_all.backward()
        optimizer.step()
        print(
            '{:04d} {:05d} Loss1={:.6f}, Lambda={}, Loss2={:.6f}, Loss_Full={:.6f}, Time={:.4f}'
            .format(epoch, iteration, np.mean(loss1.item()), Lambda,
                    np.mean(loss2.item()), np.mean(loss_all.item()),
                    time.time() - st))

    scheduler.step()

    if epoch % opt.n_snapshot == 0 or epoch == opt.n_epoch:
        network.eval()
        # save checkpoint
        checkpoint(network, epoch, "model")
        # validation
        save_model_path = os.path.join(opt.save_model_path, opt.log_name,
                                       systime)
        validation_path = os.path.join(save_model_path, "validation")
        os.makedirs(validation_path, exist_ok=True)
        np.random.seed(101)

        for valid_name, valid_data in valid_dict.items():
            psnr_result = []
            ssim_result = []
            num_img, num_block, valid_noisy, valid_gt = valid_data
            for idx in range(num_img):
                for idy in range(num_block):
                    im = valid_gt[idx, idy][:, :, np.newaxis]
                    noisy_im = valid_noisy[idx, idy][:, :, np.newaxis]

                    origin255 = im.copy() * 255.0
                    origin255 = origin255.astype(np.uint8)
                    noisy255 = noisy_im.copy() * 255.0
                    noisy255 = noisy255.astype(np.uint8)
                    # padding to square
                    H = noisy_im.shape[0]
                    W = noisy_im.shape[1]
                    val_size = (max(H, W) + 31) // 32 * 32
                    noisy_im = np.pad(
                        noisy_im,
                        [[0, val_size - H], [0, val_size - W], [0, 0]],
                        'reflect')
                    transformer = transforms.Compose([transforms.ToTensor()])
                    noisy_im = transformer(noisy_im)
                    noisy_im = torch.unsqueeze(noisy_im, 0)
                    noisy_im = noisy_im.cuda()
                    # pack raw data
                    noisy_im = space_to_depth(noisy_im, block_size=2)
                    with torch.no_grad():
                        prediction = network(noisy_im)
                        # unpack raw data
                        prediction = depth_to_space(prediction, block_size=2)
                        prediction = prediction[:, :, :H, :W]
                    prediction = prediction.permute(0, 2, 3, 1)
                    prediction = prediction.cpu().data.clamp(0, 1).numpy()
                    prediction = prediction.squeeze(0)
                    pred255 = np.clip(prediction * 255.0 + 0.5, 0,
                                        255).astype(np.uint8)
                    # calculate psnr
                    cur_psnr = calculate_psnr(origin255.astype(np.float32),
                                                pred255.astype(np.float32))
                    psnr_result.append(cur_psnr)
                    cur_ssim = calculate_ssim(origin255.astype(np.float32),
                                                pred255.astype(np.float32))
                    ssim_result.append(cur_ssim)

                    # visualization
                    save_path = os.path.join(
                        validation_path,
                        "{}_{:03d}-{:03d}-{:03d}_clean.png".format(
                            valid_name, idx, idy, epoch))
                    Image.fromarray(origin255.squeeze()).save(save_path)
                    save_path = os.path.join(
                        validation_path,
                        "{}_{:03d}-{:03d}-{:03d}_noisy.png".format(
                            valid_name, idx, idy, epoch))
                    Image.fromarray(noisy255.squeeze()).save(save_path)

                    save_path = os.path.join(
                        validation_path,
                        "{}_{:03d}-{:03d}-{:03d}_denoised.png".format(
                            valid_name, idx, idy, epoch))
                    Image.fromarray(pred255.squeeze()).save(save_path)

            psnr_result = np.array(psnr_result)
            avg_psnr = np.mean(psnr_result)
            avg_ssim = np.mean(ssim_result)
            log_path = os.path.join(validation_path,
                                    "A_log_{}.csv".format(valid_name))
            with open(log_path, "a") as f:
                f.writelines("{},{},{}\n".format(epoch, avg_psnr, avg_ssim))
